drivers/i2c/i2c_sam_twihs_rtio.c

```diff
@@ -0,0 +1,450 @@
+				i2c_sam_twihs_complete(dev, 0);
+				return;
+			}
+		} else {
+			twihs->TWIHS_THR = sqe->buf[dev_data->buf_idx++];
+		}
+	}
+
+	/* Transmission completed */
+	if (isr_status & TWIHS_SR_TXCOMP) {
+		i2c_sam_twihs_complete(dev, 0);
+	}
+}
+
+static int i2c_sam_twihs_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+				  uint16_t addr)
+{
+	struct i2c_sam_twihs_dev_data *const dev_data = dev->data;
+	struct rtio_iodev *iodev = &dev_data->iodev;
+	struct rtio *const r = dev_data->r;
+	struct rtio_sqe *sqe = NULL;
+	struct rtio_cqe *cqe = NULL;
+	int res = 0;
+
+	k_sem_take(&dev_data->block_lock, K_FOREVER);
+
+	dev_data->dt_spec.addr = addr;
+
+	sqe = i2c_rtio_copy(r, iodev, msgs, num_msgs);
+	if (sqe == NULL) {
+		LOG_ERR("Not enough submission queue entries");
+		res = -ENOMEM;
+		goto out;
+	}
+
+	sqe->flags &= ~RTIO_SQE_TRANSACTION;
+
+	rtio_submit(r, 1);
+
+	cqe = rtio_cqe_consume(r);
+	__ASSERT(cqe != NULL, "Expected valid completion");
+	while (cqe != NULL) {
+		res = cqe->result;
+		cqe = rtio_cqe_consume(r);
+	}
+	rtio_cqe_release_all(r);
+
+out:
+	k_sem_give(&dev_data->block_lock);
+	return res;
+}
+
+static int i2c_sam_twihs_initialize(const struct device *dev)
+{
+	const struct i2c_sam_twihs_dev_cfg *const dev_cfg = dev->config;
+	struct i2c_sam_twihs_dev_data *const dev_data = dev->data;
+	Twihs *const twihs = dev_cfg->regs;
+	uint32_t bitrate_cfg;
+	int ret;
+
+	/* Configure interrupts */
+	dev_cfg->irq_config();
+
+	/* Connect pins to the peripheral */
+	ret = pinctrl_apply_state(dev_cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Enable TWIHS clock in PMC */
+	(void)clock_control_on(SAM_DT_PMC_CONTROLLER,
+			       (clock_control_subsys_t *)&dev_cfg->clock_cfg);
+
+	/* Reset the module */
+	twihs->TWIHS_CR = TWIHS_CR_SWRST;
+
+	bitrate_cfg = i2c_map_dt_bitrate(dev_cfg->bitrate);
+
+	ret = i2c_sam_twihs_configure(dev, I2C_MODE_CONTROLLER | bitrate_cfg);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize %s device", dev->name);
+		return ret;
+	}
+
+	k_sem_init(&dev_data->block_lock, 1, 1);
+	dev_data->dt_spec.bus = dev;
+	dev_data->iodev.api = &i2c_iodev_api;
+	dev_data->iodev.data = &dev_data->dt_spec;
+	rtio_mpsc_init(&dev_data->io_q);
+
+	/* Enable module's IRQ */
+	irq_enable(dev_cfg->irq_id);
+
+	LOG_INF("Device %s initialized", dev->name);
+
+	return 0;
+}
+
+static const struct i2c_driver_api i2c_sam_twihs_driver_api = {
+	.configure = i2c_sam_twihs_configure,
+	.transfer = i2c_sam_twihs_transfer,
+	.iodev_submit = i2c_sam_twihs_submit,
+};
+
+#define I2C_TWIHS_SAM_INIT(n)									\
+	PINCTRL_DT_INST_DEFINE(n);								\
+	static void i2c##n##_sam_irq_config(void)						\
+	{											\
+		IRQ_CONNECT(DT_INST_IRQN(n), DT_INST_IRQ(n, priority),				\
+			    i2c_sam_twihs_isr,							\
+			    DEVICE_DT_INST_GET(n), 0);						\
+	}											\
+												\
+	RTIO_EXECUTOR_SIMPLE_DEFINE(_i2c##n##_sam_rtio_exec);					\
+	RTIO_DEFINE(_i2c##n##_sam_rtio,								\
```

Comment:
just to clarify if i understood correctly. This RTIO is only for synchronous `transfer` call. If user wants to use asynchronous API then own RTIO can be used and `submit` API?

Reply:
You got it